text/0000-proc-macros.md

```diff
@@ -0,0 +1,452 @@
+```
+
+### Open question: `Punctuation(char)` and multi-char operators.
+
+Rust has many compound operators, e.g., `<<`. It's not clear how best to deal
+with them. If the source code contains "`+ =`", it would be nice to distinguish
+this in the token stream from "`+=`". On the other hand, if we represent `<<` as
+a single token, then the macro may need to split them into `<`, `<` in generic
+position.
+
+I had hoped to represent each character as a separate token. However, to make
+pattern matching backwards compatible, we would need to combine some tokens. In
+fact, if we want to be completely backwards compatible, we probably need to keep
+the same set of compound operators as are defined at the moment.
+
+Some solutions:
+
+* `Punctuation(char)` with special rules for pattern matching tokens,
+* `Punctuation([char])` with a facility for macros to split tokens. Tokenising
+  could match the maximum number of punctuation characters, or use the rules for
+  the current token set. The former would have issues with pattern matching. The
+  latter is a bit hacky, there would be backwards compatibility issues if we
+  wanted to add new compound operators in the future.
+
+## Staging
+
+1. Implement [RFC 1561](https://github.com/rust-lang/rfcs/pull/1561).
+2. Implement `#[macro]` and `#[cfg(macro)]` and the function approach to
+   defining macros. However, pass the existing data structures to the macros,
+   rather than tokens and `MacroContext`.
+3. Implement libmacro and make this available to macros. At this stage both old
+   and new macros are available (functions with different signatures). This will
+   require an RFC and considerable refactoring of the compiler.
+4. Implement some high-level macro facilities in external crates on top of
+   libmacro. It is hoped that much of this work will be community-led.
+5. After some time to allow conversion, deprecate the old-style macros. Later,
+   remove old macros completely.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Procedural macros are a somewhat unpleasant corner of Rust at the moment. It is
+hard to argue that some kind of reform is unnecessary. One could find fault with
+this proposed reform in particular (see below for some alternatives). Some
+drawbacks that come to mind:
+
+* providing such a low-level API risks never seeing good high-level libraries;
+* the design is complex and thus will take some time to implement and stabilise,
+  meanwhile unstable procedural macros are a major pain point in current Rust;
+* dealing with tokens and hygiene may discourage macro authors due to complexity,
+  hopefully that is addressed by library crates.
+
+The actual concept of procedural macros also have drawbacks: executing arbitrary
+code in the compiler makes it vulnerable to crashes and possibly security issues,
+macros can introduce hard to debug errors, macros can make a program hard to
+comprehend, it risks creating de facto dialects of Rust and thus fragmentation
```

Comment:
Has the problem with dialects already come up in your experience with macros 1.0? What do you plan to do if this ends up being an issue?